### PR TITLE
Refactor metrics and metric tables and support adding more in-training metrics.

### PR DESCRIPTION
**Pull request by justinxzhao**
Changes:
- Rename `metric_type_registry` -> `metric_feature_type_registry`
- Add Precision and Recall metrics.
- Remove `get_output_metric_functions()` everywhere and rely on the metric registry to determine which metrics correspond to which feature types.
- For output features, make `metric_functions` private, in favor of a public `metric_names` attribute. Metrics are setup, updated, and managed internally within the `OutputFeature` -- we don't need to the expose the actual functions.
- Use the metric registry to populate options for the hyperopt schema.
- Clean up metric table printing mechanisms by refactoring metric tables to be constructed outside of `trainer.evaluation()` (instead of appended to during evaluation). Metrics printing is factored out into a separate class, `MetricsPrintedTable`.
- Add accessor methods for metrics-related registries, and use them.

### CODE REVIEW

File: ludwig/modules/metric_modules.py [hunk 1]
```diff
@@ -129,75 +132,44 @@ def get_inputs(cls):
...
-        false_positives = self.summary_stats[:, 0]
-        true_negatives = self.summary_stats[:, 1]
-        false_negatives = self.summary_stats[:, 2]
-        true_positives = self.summary_stats[:, 3]
 
-        true_positive_rate = true_positives / (true_positives + false_negatives)
-        false_positive_rate = false_positives / (false_positives + true_negatives)
+@register_metric(ROC_AUC, [BINARY, CATEGORY])
+class AUROCMetric(AUROC, LudwigMetric):
+    """Recall metric."""
```
Comment (dantreiman) [inline ludwig/modules/metric_modules.py:169]:
```suggestion
    """Area under the receiver operating curve."""
```

File: ludwig/constants.py [hunk 2]
```diff
@@ -53,7 +53,7 @@
 SEQUENCE_SOFTMAX_CROSS_ENTROPY = "sequence_softmax_cross_entropy"
 SOFTMAX_CROSS_ENTROPY = "softmax_cross_entropy"
 SIGMOID_CROSS_ENTROPY = "sigmoid_cross_entropy"
-BINARY_WEIGHTED_CROSS_ENTROPY = "binary_weighted_cross_entropy"
+BINARY_WEIGHTED_CROSS_ENTROPY_LOSS = "binary_weighted_cross_entropy_loss"
```
Comment (dantreiman) [inline ludwig/constants.py:56]: Why is `BINARY_WEIGHTED_CROSS_ENTROPY` suffixed with `_LOSS`, when none of the other cross_entropy metrics are (i.e. `SOFTMAX_CROSS_ENTROPY`, `SIGMOID_CROSS_ENTROPY`)?

File: ludwig/schema/hyperopt/__init__.py [hunk 3]
```diff
@@ -5,8 +5,9 @@
 
 from ludwig.api_annotations import DeveloperAPI
 from ludwig.constants import LOSS, TEST, TRAIN, VALIDATION
+from ludwig.modules import metric_modules  # noqa: Needed to ensure that the metric registry is populated.
+from ludwig.modules.metric_registry import metric_registry
```
Comment (dantreiman) [inline ludwig/schema/hyperopt/__init__.py:9]:
```suggestion
from ludwig.modules.metric_registry import get_metric_registry
```

File: ludwig/schema/hyperopt/__init__.py [hunk 4]
```diff
@@ -36,7 +30,7 @@ def get_hyperopt_metric_options():
     )
 
     metric: str = schema_utils.StringOptions(
-        options=get_hyperopt_metric_options(),
+        options=metric_registry.keys(),
```
Comment (dantreiman) [inline ludwig/schema/hyperopt/__init__.py:33]:
```suggestion
        options=get_metric_registry().keys(),
```

File: ludwig/utils/trainer_utils.py [hunk 5]
```diff
@@ -244,23 +256,49 @@ def log_metrics(self):
         return log_metrics
 
 
+def add_metrics_to_printed_table(
+    printed_table: List,
```
Comment (dantreiman) [inline ludwig/utils/trainer_utils.py:260]:
```suggestion
    printed_table: Dict[str, List],
```
Also consider above suggestion to turn this into a class.

**Review by dantreiman — CHANGES_REQUESTED** [review]

File: ludwig/utils/metric_utils.py [hunk 6]
```diff
@@ -89,13 +91,24 @@ def reduce_trainer_metrics_dict(
...
+    return metrics_names
+
+
+def get_feature_to_metric_names_map(output_features: List[FeatureConfigDict]) -> Dict[str, List[str]]:
+    """Returns a dict of output_feature_name -> list of metric names."""
+    metrics_names = {}
+    for output_feature in output_features:
+        output_feature_name = output_feature[NAME]
+        output_feature_type = output_feature[TYPE]
+        metrics_names[output_feature_name] = get_metric_feature_type_registry()[output_feature_type]
```
Comment (connor-mccorm) [inline ludwig/utils/metric_utils.py:112]: Suggestion: A lot of the other registries have accessor methods (i.e. `get_encoder_classes`). It may be worthwhile to add a similar accessor method here for the `metric_feature_type_registry` to remain consistent and make it a little clearer what you're actually getting on this line of code.

File: ludwig/utils/trainer_utils.py [hunk 7]
```diff
@@ -244,23 +256,49 @@ def log_metrics(self):
...
+        # [0]: The header is the first row, which contains names of metrics.
+        # [1:]: Skip the first column as it's just the name of the output feature, not an actual metric name.
+        for metric_name in printed_table[output_feature_name][0][1:]:
+            # Metrics may be missing if should_evaluate_train is False.
+            if metric_name in output_feature_metrics and output_feature_metrics[metric_name]:
+                printed_metrics.append(output_feature_metrics[metric_name][-1][-1])
+            else:
+                printed_metrics.append("")
+
+        # The printed table.
```
Comment (connor-mccorm) [inline ludwig/utils/trainer_utils.py:276]: Love this!

**Review by connor-mccorm — APPROVED** [review]
Looks great! Just a couple small comments but besides that I'm good to go on this. Thanks for putting this together:)

File: ludwig/utils/trainer_utils.py [hunk 8]
```diff
@@ -18,14 +18,26 @@
 logger = logging.getLogger(__name__)
 
 
+@DeveloperAPI
+def initialize_printed_table(
```
[the remark this answers is not in the document]
Comment (justinxzhao) [reply ludwig/utils/trainer_utils.py]: Great suggestion! Made a new class and moved it to its own file.

**Review by dantreiman — APPROVED** [review]
